File: PetVue/src/utils/ratingCalculator.ts
/**
 * 评分计算工具
 * 
 * 实现评价系统的评分计算逻辑
 */

import type { Review, ReviewRating } from '@/types/review'

/**
 * 计算综合平均评分
 */
export function calculateAverageRating(reviews: Review[]): number {
  if (reviews.length === 0) return 0
  
  const sum = reviews.reduce((acc, review) => acc + review.rating.overall, 0)
  return Math.round((sum / reviews.length) * 10) / 10
}

/**
 * 计算各维度平均评分
 */
export function calculateDimensionAverages(reviews: Review[]): ReviewRating {
  if (reviews.length === 0) {
    return { overall: 0, environment: 0, service: 0, hygiene: 0, communication: 0 }
  }
  
  const dimensions: (keyof ReviewRating)[] = ['overall', 'environment', 'service', 'hygiene', 'communication']
  const result: ReviewRating = { overall: 0, environment: 0, service: 0, hygiene: 0, communication: 0 }
  
  for (const dim of dimensions) {
    const sum = reviews.reduce((acc, review) => acc + review.rating[dim], 0)
    result[dim] = Math.round((sum / reviews.length) * 10) / 10
  }
  
  return result
}

/**
 * 计算评分分布
 */
export function calculateRatingDistribution(reviews: Review[]): Record<number, number> {
  const distribution: Record<number, number> = { 1: 0, 2: 0, 3: 0, 4: 0, 5: 0 }
  
  reviews.forEach(review => {
    const rating = Math.round(review.rating.overall)
    if (rating >= 1 && rating <= 5) {
      distribution[rating]++
    }
  })
  
  return distribution
}

/**
 * 计算评分分布百分比
 */
export function calculateRatingPercentages(reviews: Review[]): Record<number, number> {
  const distribution = calculateRatingDistribution(reviews)
  const total = reviews.length
  
  if (total === 0) {
    return { 1: 0, 2: 0, 3: 0, 4: 0, 5: 0 }
  }
  
  const percentages: Record<number, number> = {}
  for (let i = 1; i <= 5; i++) {
    percentages[i] = Math.round((distribution[i] / total) * 100)
  }
  
  return percentages
}

/**
 * 获取评分等级描述
 */
export function getRatingLevel(rating: number): { level: string; color: string } {
  if (rating >= 4.5) return { level: '极好', color: '#52c41a' }
  if (rating >= 4.0) return { level: '很好', color: '#73d13d' }
  if (rating >= 3.5) return { level: '较好', color: '#faad14' }
  if (rating >= 3.0) return { level: '一般', color: '#fa8c16' }
  if (rating >= 2.0) return { level: '较差', color: '#ff7a45' }
  return { level: '很差', color: '#ff4d4f' }
}

/**
 * 格式化评分显示
 */
export function formatRating(rating: number): string {
  return rating.toFixed(1)
}

/**
 * 验证评价内容长度
 */
export function validateReviewContent(
  content: string, 
  minLength = 10, 
  maxLength = 500
): { valid: boolean; error?: string } {
  const trimmed = content.trim()
  
  if (trimmed.length < minLength) {
    return { 
      valid: false, 
      error: `评价内容至少需要 ${minLength} 个字符` 
    }
  }
  
  if (trimmed.length > maxLength) {
    return { 
      valid: false, 
      error: `评价内容不能超过 ${maxLength} 个字符` 
    }
  }
  
  return { valid: true }
}

/**
 * 计算推荐指数
 * 基于评分和评价数量
 */
export function calculateRecommendationScore(
  averageRating: number, 
  reviewCount: number
): number {
  if (reviewCount === 0) return 0
  
  // 贝叶斯平均，考虑评价数量的影响
  const C = 10 // 置信度参数
  const m = 3.5 // 全局平均评分
  
  const score = (reviewCount * averageRating + C * m) / (reviewCount + C)
  return Math.round(score * 10) / 10
}

File: PetJava/src/main/java/com/pet/util/RatingCalculator.java
package com.pet.util;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 评分计算工具
 */
public class RatingCalculator {

    /**
     * 计算综合平均评分
     */
    public static BigDecimal calculateAverageRating(List<Map<String, Object>> reviews) {
        if (reviews == null || reviews.isEmpty()) {
            return BigDecimal.ZERO;
        }

        BigDecimal sum = BigDecimal.ZERO;
        for (Map<String, Object> review : reviews) {
            Object ratingObj = review.get("rating");
            if (ratingObj instanceof Map) {
                @SuppressWarnings("unchecked")
                Map<String, Object> rating = (Map<String, Object>) ratingObj;
                Object overall = rating.get("overall");
                if (overall instanceof Number) {
                    sum = sum.add(BigDecimal.valueOf(((Number) overall).doubleValue()));
                }
            }
        }

        return sum.divide(BigDecimal.valueOf(reviews.size()), 1, RoundingMode.HALF_UP);
    }

    /**
     * 计算各维度平均评分
     */
    public static Map<String, BigDecimal> calculateDimensionAverages(List<Map<String, Object>> reviews) {
        Map<String, BigDecimal> result = new HashMap<>();
        String[] dimensions = {"overall", "environment", "service", "hygiene", "communication"};

        for (String dim : dimensions) {
            result.put(dim, BigDecimal.ZERO);
        }

        if (reviews == null || reviews.isEmpty()) {
            return result;
        }

        Map<String, BigDecimal> sums = new HashMap<>();
        for (String dim : dimensions) {
            sums.put(dim, BigDecimal.ZERO);
        }

        for (Map<String, Object> review : reviews) {
            Object ratingObj = review.get("rating");
            if (ratingObj instanceof Map) {
                @SuppressWarnings("unchecked")
                Map<String, Object> rating = (Map<String, Object>) ratingObj;
                for (String dim : dimensions) {
                    Object value = rating.get(dim);
                    if (value instanceof Number) {
                        sums.put(dim, sums.get(dim).add(BigDecimal.valueOf(((Number) value).doubleValue())));
                    }
                }
            }
        }

        int count = reviews.size();
        for (String dim : dimensions) {
            result.put(dim, sums.get(dim).divide(BigDecimal.valueOf(count), 1, RoundingMode.HALF_UP));
        }

        return result;
    }

    /**
     * 计算评分分布
     */
    public static Map<Integer, Integer> calculateRatingDistribution(List<Map<String, Object>> reviews) {
        Map<Integer, Integer> distribution = new HashMap<>();
        for (int i = 1; i <= 5; i++) {
            distribution.put(i, 0);
        }

        if (reviews == null || reviews.isEmpty()) {
            return distribution;
        }

        for (Map<String, Object> review : reviews) {
            Object ratingObj = review.get("rating");
            if (ratingObj instanceof Map) {
                @SuppressWarnings("unchecked")
                Map<String, Object> rating = (Map<String, Object>) ratingObj;
                Object overall = rating.get("overall");
                if (overall instanceof Number) {
                    int rounded = (int) Math.round(((Number) overall).doubleValue());
                    if (rounded >= 1 && rounded <= 5) {
                        distribution.put(rounded, distribution.get(rounded) + 1);
                    }
                }
            }
        }

        return distribution;
    }

    /**
     * 获取评分等级描述
     */
    public static Map<String, String> getRatingLevel(double rating) {
        Map<String, String> result = new HashMap<>();
        if (rating >= 4.5) {
            result.put("level", "极好");
            result.put("color", "#52c41a");
        } else if (rating >= 4.0) {
            result.put("level", "很好");
            result.put("color", "#73d13d");
        } else if (rating >= 3.5) {
            result.put("level", "较好");
            result.put("color", "#faad14");
        } else if (rating >= 3.0) {
            result.put("level", "一般");
            result.put("color", "#fa8c16");
        } else if (rating >= 2.0) {
            result.put("level", "较差");
            result.put("color", "#ff7a45");
        } else {
            result.put("level", "很差");
            result.put("color", "#ff4d4f");
        }
        return result;
    }

    /**
     * 验证评价内容长度
     */
    public static Map<String, Object> validateReviewContent(String content, int minLength, int maxLength) {
        Map<String, Object> result = new HashMap<>();
        String trimmed = content != null ? content.trim() : "";

        if (trimmed.length() < minLength) {
            result.put("valid", false);
            result.put("error", "评价内容至少需要 " + minLength + " 个字符");
            return result;
        }

        if (trimmed.length() > maxLength) {
            result.put("valid", false);
            result.put("error", "评价内容不能超过 " + maxLength + " 个字符");
            return result;
        }

        result.put("valid", true);
        return result;
    }

    /**
     * 计算推荐指数（贝叶斯平均）
     */
    public static BigDecimal calculateRecommendationScore(double averageRating, int reviewCount) {
        if (reviewCount == 0) {
            return BigDecimal.ZERO;
        }

        double C = 10; // 置信度参数
        double m = 3.5; // 全局平均评分

        double score = (reviewCount * averageRating + C * m) / (reviewCount + C);
        return BigDecimal.valueOf(score).setScale(1, RoundingMode.HALF_UP);
    }
}
